core/src/test/java/org/infinispan/conflict/impl/ConflictManagerTest.java

```diff
@@ -74,61 +75,48 @@ public void testGetAllVersionsDuringStateTransfer() throws Exception {
       createCluster();
       getCache(2).put(key, value);
       splitCluster();
-
-      delayStateTransferCompletion();
       RehashListener listener = new RehashListener();
-      getCache(2).addListener(listener);
+      getCache(0).addListener(listener);
+      CountDownLatch latch = new CountDownLatch(1);
+      delayStateTransferCompletion(latch);
 
       Future<Map<Address, InternalCacheValue<Object>>> versionFuture = fork(() -> {
-         while (!listener.isRehashInProgress.get())
-            TestingUtil.sleepThread(100);
-
+         listener.latch.await();
+         fork(latch::countDown);
```

Comment:
Actually I think the count-down doesn't belong here. I'd change the test to something like this:

``` 
      // Trigger the merge and wait for state transfer to begin
      Future<?> mergeFuture = fork(() -> partition(0).merge(partition(1)));
      listener.latch.await();

      Future<Map<Address, InternalCacheValue<Object>>> versionFuture = fork(() -> getAllVersions(0, key));

      // Check that getAllVersions doesn't return while state transfer is in progress
      Exceptions.expectException(TimeoutException.class, () -> versionFuture.get(500, TimeUnit.MILLISECONDS));

      // Allow and wait for state transfer to finish
      latch.countDown();
      mergeFuture.get(30, TimeUnit.SECONDS);

      // Check the results
      Map<Address, InternalCacheValue<Object>> versionMap = versionFuture.get(10, TimeUnit.SECONDS);
      assertTrue(versionMap != null);
      assertTrue(!versionMap.isEmpty());
      assertEquals(String.format("Returned versionMap %s", versionMap),2, versionMap.size());
      versionMap.values().forEach(icv -> assertEquals(value, icv.getValue()));
```